Sync de registration sharing when removing a system

jesusbv opened the pull request:
Sync de registration to other RMT servers, if any,
when removing a system from within RMT using the cli

This fixes bsc#1185011

## Description

When running `rmt-cli systems remove XXX` and there are other peers to share that
removal of the registration, it did not happen
This adds that sharing of the removal of that system to the other peers

Fixes bsc#1185011

## Change Type

*Please select the correct option.*

- [X] **Bug Fix** (a non-breaking change which fixes an issue)
- [ ] **New Feature** (a non-breaking change which adds new functionality)
- [ ] **Documentation Update** (a change which only updates documentation)

## Checklist

*Please check off each item if the requirement is met.*

- [X] I have verified that my code follows RMT's coding standards with `rubocop`.
- [X] I have reviewed my own code and believe that it's ready for an external review.
- [X] I have provided comments for any hard-to-understand code.
- [ ] I have documented the `MANUAL.md` file with any changes to the user experience.
- [X] RMT's test coverage remains at 100%.
- [ ] If my changes are non-trivial, I have added a changelog entry to notify users at `package/obs/rmt-server.changes`.

## Other Notes

To demo the changes, run two RMT instances configured to be synced (with regsharing)
and another instance registering to any of them, e.g.
```bash
sudo registercloudguest --force-new --smt-ip rmt_instance_ip --smtfqdn......`
```
Check that the found_registered_system_login is present in both RMT instances
e.g.
```bash
rmt-cli systems list -a | grep found_registered_system_login`
```

Then, inside one of the RMT instances, get the new system login for that registration, and
run `rmt-cli systems remove found_registered_system_login`

Check that the found_registered_system_login is *not* present in both RMT instances
e.g.
```bash
rmt-cli systems list -a | grep found_registered_system_login`
```

suse-tests-pass commented:
<!--
  0 Errors
  1 Warning: Unless this is a trivial chang...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Unless this is a trivial change, please include a CHANGELOG entry.<br />
Run <code>osc vc</code> in the <code>package</code> directory to add one.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

digitaltom commented:
Would it be possible to move this code into the `registration_sharing` engine? Because it would only be used when this engine is installed.

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/192947/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged a791a1410f5c7b139545df2b55d6b85971961e9c into 6da51f61d7bf2a6cea0e9b937222c795da9cbb9a)

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/193266/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged ac4422a1e770c63ca5dbd60020c22818ba209ed6 into 6da51f61d7bf2a6cea0e9b937222c795da9cbb9a)

jesusbv commented:
> Would it be possible to move this code into the `registration_sharing` engine? Because it would only be used when this engine is installed.

Hi @digitaltom , I'm debugging why it is not triggered so there is no need for this code outside `registration_sharing`
Given that in bsc#1185011 there is a customer waiting, a fix for that behavior and keep investigating the issue seems reasonable, thoughts ?

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/193270/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 3ef31c6325cba50f58c6a1acaa6aec62c2e25aa4 into 6da51f61d7bf2a6cea0e9b937222c795da9cbb9a)

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/193772/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 3ef31c6325cba50f58c6a1acaa6aec62c2e25aa4 into 55ce6a2480f8c17518611a3adc3b27a384ba0e40)

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/193906/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 3ef31c6325cba50f58c6a1acaa6aec62c2e25aa4 into 71386fd62b5658eb00ae2881b8a9b8b12ae1e83c)

digitaltom commented:
Hi @jesusbv did you succeed implementing this in the `registration_sharing` engine? If you need more help, maybe someone from the team can pair with you. You can reach us anytime in #scc-help. 
The point of having this specific code under `engines/registration_sharing` has 2 reasons: 

* It belongs there from the technical architecture, because this is an optional feature which even gets delivered in another rpm package. 
* The `engines/registration_sharing` code and the other optional engines are owned and maintained by the public cloud team. We (SCC Team) don't do an in-depth review on those, because we don't maintain them. That's an advantage for you, because you don't need our ok for the code, and you can review and release this based on your team workflows without being blocked by us. The code that is part of core RMT needs to be reviewed, understood and maintained by the SCC team, and also needs to get presented in our Demo meeting.

suse-tests-fail commented:
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/193939/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests. (Merged 3ef31c6325cba50f58c6a1acaa6aec62c2e25aa4 into 177d406d9bef02b09b73ad13731a94aa2c20ef6a)

suse-tests-fail commented:
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/193940/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests.
Failed fast in project rmt on step docker-compose-run-tests.
Failed fast in project rmt on step docker-compose-run-tests. (Merged 3ef31c6325cba50f58c6a1acaa6aec62c2e25aa4 into 43c9eed40cfe99dd2456bcf4d185671c153e5127)

jesusbv commented:
Hi @digitaltom 

> Hi @jesusbv did you succeed implementing this in the `registration_sharing` engine? 
I do not think the registration sharing can be triggered from the rmt-cli systems remove command 
with code only in `registration_sharing`, without editing `rmt-cli systems remove` command.

I have updated the PR with code that works (as in it deletes the system in other peers) and
that code won't get triggered/invoked/loaded if no regsharing is set in the server.

> The point of having this specific code under `engines/registration_sharing` has 2 reasons:
> 
>     * It belongs there from the technical architecture, because this is an optional feature which even gets delivered in another rpm package.
>
I get that.
 
>     * The `engines/registration_sharing` code and the other optional engines are owned and maintained by the public cloud team. We (SCC Team) don't do an in-depth review on those, because we don't maintain them. That's an advantage for you, because you don't need our ok for the code, and you can review and release this based on your team workflows without being blocked by us. The code that is part of core RMT needs to be reviewed, understood and maintained by the SCC team, and also needs to get presented in our Demo meeting.

I get that, that's why this latest update on the PR does not create any new code, all the code is already in use, I added
a line in the cli tool to re use the code already implemented and working in `engines/registration_sharing` conditionally.

Please, let me know your thoughts/feedback.

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/194234/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 25ae509739cb898836220d3528427777f8f5ae41 into 43c9eed40cfe99dd2456bcf4d185671c153e5127)

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/194266/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 0ef41dd3524ab6019cd1d21c6495f52f5e416d88 into 43c9eed40cfe99dd2456bcf4d185671c153e5127)

digitaltom commented on spec/lib/rmt/cli/systems_spec.rb:
```diff
@@ -133,6 +133,16 @@
+        it 'de registration is shared with peers' do
+          allow(Settings).to receive(:[]).with(:regsharing).and_return({ peers: 'foo' })
+
+          expect { described_class.start(argv) }
```
This test should actually check if `RegistrationSharing.save_for_sharing` gets called correctly.

digitaltom commented:
I see it's hard to use the model overrides that are already defined in the engine (`engines/registration_sharing/lib/registration_sharing/engine.rb:14`) from within the `rmt-cli` script. If you find out a way it would be very cool, because the code to share the registration on destroy is already there. 

But so far we can go with this, please just check my comment on the test.

jesusbv commented:
> I see it's hard to use the model overrides that are already defined in the engine (`engines/registration_sharing/lib/registration_sharing/engine.rb:14`) from within the `rmt-cli` script. If you find out a way it would be very cool, because the code to share the registration on destroy is already there.
> 

Yes, I have not found a way (yet) to do it with that approach, which was my original intention as well. Given it is somewhat time sensitive (a bug is open), thought best to fix it as good as possible, less `registration_sharing` code in `rmt-cli` land as possible.

> But so far we can go with this

Sounds good to me

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/194291/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged e50e16bccf14e921d88fc48b285a737ed0cc688f into 43c9eed40cfe99dd2456bcf4d185671c153e5127)

digitaltom reviewed (approved):
LGTM with regards to the RMT core part. I cannot test and verify if the system removal gets propagated to the other RMTs that are connected via `registration_sharing`, because we have no experience how to set that up. 

Regarding the SCC Review meeting: I suggest that either I demo that the normal RMT functionality is not impacted by this PR, or you could demo to the team the whole process as described in the PR description. What do you think?

jesusbv commented:
> Regarding the SCC Review meeting: I suggest that either I demo that the normal RMT functionality is not impacted by this PR, or you could demo to the team the whole process as described in the PR description. What do you think?

I don't mind, no strong preference, maybe you guys are interested in seeing the registration sharing ?

digitaltom commented:
> I don't mind, no strong preference, maybe you guys are interested in seeing the registration sharing ?

Sure, that would be great if you could show us how that works. We have the next Demo meeting on Tuesday 25th at 10:00 CET. Can I invite you there to show it?

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/194401/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged e50e16bccf14e921d88fc48b285a737ed0cc688f into 81b807b61c9aa283f10d80dd7a52d2dd36543ad9)

suse-tests-pass commented:
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/194402/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 8656a63403c480958cbf1d2736a7078a0fa884e5 into 81b807b61c9aa283f10d80dd7a52d2dd36543ad9)
